src/libraries/Common/src/System/Net/Http/aspnetcore/Quic/Implementations/Managed/Internal/Buffers/InboundBuffer.cs

```diff
@@ -0,0 +1,253 @@
+
+        /// <summary>
+        ///     Total number of bytes allowed to transport in this stream.
+        /// </summary>
+        internal long MaxData { get; private set; }
+
+        /// <summary>
+        ///     Updates the <see cref="MaxData"/> parameter.
+        /// </summary>
+        /// <param name="value">Value of the parameter.</param>
+        internal void UpdateMaxData(long value)
+        {
+            MaxData = Math.Max(MaxData, value);
+        }
+
+        public InboundBuffer(long maxData)
+        {
+            UpdateMaxData(maxData);
+        }
+
+        /// <summary>
+        ///     Ranges of data which are received, but not delivered.
+        /// </summary>
+        private readonly RangeSet _undelivered = new RangeSet();
+
+        /// <summary>
+        ///     Number of bytes streamed through the <see cref="_deliverableChannel"/>.
+        /// </summary>
+        internal long _bytesChanneled;
+
+        /// <summary>
+        ///     Total number of bytes delivered.
+        /// </summary>
+        internal long BytesRead { get; private set; }
+
+        /// <summary>
+        ///     Final size of the stream. Null if final size is not known yet.
+        /// </summary>
+        internal long? FinalSize { get; set; }
+
+        /// <summary>
+        ///     Estimated size of the stream. Final size may not be lower than this value.
+        /// </summary>
+        internal long EstimatedSize => _undelivered.Count > 0 ? _undelivered.GetMax() : BytesRead;
+
+        /// <summary>
+        ///     Number of bytes ready to be read from the stream.
+        /// </summary>
+        internal long BytesAvailable => _bytesChanneled - BytesRead;
+
+        /// <summary>
+        ///     Receives a chunk of data and buffers it for delivery.
+        /// </summary>
+        /// <param name="offset">Offset on the stream of the received data.</param>
+        /// <param name="data">The data to be received.</param>
+        internal void Receive(long offset, ReadOnlySpan<byte> data)
+        {
+            Debug.Assert(FinalSize == null || offset + data.Length <= FinalSize, "Writing after final size");
+
+            if (data.IsEmpty)
+            {
+                return;
+            }
+
+            long deliverable = BytesRead + BytesAvailable;
+            if (offset + data.Length <= deliverable)
+            {
+                return; // entirely duplicate
+            }
+
+            if (offset < deliverable)
+            {
+                // drop duplicate prefix;
+                data = data.Slice((int)(deliverable - offset));
+                offset = deliverable;
+            }
+
+
+            // optimized hot path - in-order delivery
+            if (_chunks.Count == 0 || _chunks[^1].StreamOffset + _chunks[^1].Length <= offset)
+            {
+                var buffer = ArrayPool<byte>.Shared.Rent(data.Length);
+                data.CopyTo(buffer);
```

Comment:
Edit: I accidentally deleted @ManickaP's comment:
> Correct me if I'm wrong, but the data come from StreamFrame data which come from QuicReader which is held in Memory buffer. So the copying here could be avoided here if Memory was passed in instead of Span.

It is not that easy to remove those temp copies, the Memory backing the reader is the buffer into which I receive UDP datagrams. For that I may need up to 64kb of contiguous space (or less, depending on the size of maximum accepted UDP datagram). if I removed this copy, I would have to wait until user reads the data before reusing the buffer, possibly leaking memory if only few bytes are left unread.

I have created an issue #6 to track ideas how to remove additional copies. I intend to focus on it more in late phase of implementation.